### PR TITLE
[Feat] Thin plate spline warping

### PR DESCRIPTION
**Pull request by catalys1**
### Description
Implements functionality for computing thin plate spline warps, as described in #872 

### Status
**Work in progress**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: test/geometry/transform/test_thin_plate_spline.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            src = torch.rand(batch_size, 5)
+            assert kornia.get_tps_transform(src, src)
+
+    @pytest.mark.grad
+    @pytest.mark.parametrize('batch_size', [1, 3])
+    @pytest.mark.parametrize('requires_grad', [True, False])
+    def test_gradcheck(self, batch_size, device, dtype, requires_grad):
+        src = torch.rand(batch_size, 5, 2, device=device, requires_grad=requires_grad)
+        dst = torch.rand(batch_size, 5, 2, device=device, requires_grad=(not requires_grad))
+        assert gradcheck(kornia.get_tps_transform, (src, dst), raise_exception=True)
```
Comment (edgarriba) [inline test/geometry/transform/test_thin_plate_spline.py:56]: seems that gradient check doesn't work properly. What I would suggest is that you perform the same tests for the intermediate functions in order to find what blocks from backpropagating and run the tests locally.

File: kornia/geometry/transform/thin_plate_spline.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    device, dtype = _extract_device_dtype([points_src, points_dst])
+    batch_size, num_points = points_src.shape[:2]
+
+    # set up and solve linear system
+    # [K   P] [w] = [dst]
+    # [P^T 0] [a]   [ 0 ]
+    pair_distance: torch.Tensor = _pair_square_euclidean(points_src, points_dst)
+    k_matrix: torch.Tensor = _kernel_distance(pair_distance)
+
+    requires_grad = points_src.requires_grad or points_dst.requires_grad
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:78]: where is this used ?

File: kornia/geometry/transform/thin_plate_spline.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    if not isinstance(points_dst, torch.Tensor):
+        raise TypeError(f"Input points_dst is not torch.Tensor. Got {type(points_dst)}")
+
+    if not len(points_src.shape) == 3:
+        raise ValueError(f"Invalid shape for points_src, expected BxNx2. Got {points_src.shape}")
+
+    if not len(points_dst.shape) == 3:
+        raise ValueError(f"Invalid shape for points_dst, expected BxNx2. Got {points_dst.shape}")
+
+    device, dtype = _extract_device_dtype([points_src, points_dst])
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:69]: this function seems to break jit. Test in local to make sure that it works. What I would suggest is to get rid of this function and get directly the device/dtype from src or dst. In case they don't match, pytorch itself will raise an error that usually is easy for the user to understand what's going on.

**Review by edgarriba — CHANGES_REQUESTED** [review]

**Comment by edgarriba** [comment]
@catalys1 just minor comments

File: kornia/geometry/transform/thin_plate_spline.py (hunk 2 above)
Comment (catalys1) [reply kornia/geometry/transform/thin_plate_spline.py:78]: I'm not sure how that got there, maybe an accidental copy/paste...

File: test/geometry/transform/test_thin_plate_spline.py (hunk 1 above)
Comment (catalys1) [reply test/geometry/transform/test_thin_plate_spline.py:57]: There were a few things I didn't realize about gradcheck, particularly that all the arguments needed to be float64.

**Comment by catalys1** [comment]
@edgarriba Fixed the issues, tests are passing on my local now.

**Review by edgarriba — APPROVED** [review]

**Comment by edgarriba** [comment]
@catalys1 done!! thanks for the contribution. Just one missing thing - add the function to the docs here:
https://github.com/kornia/kornia/blob/master/docs/source/geometry.transform.rst

**Comment by edgarriba** [comment]
@catalys1 do you have some a notebook with images to include in the release notes ? or potentially as a tutorial here ?
https://github.com/kornia/tutorials

**Comment by catalys1** [comment]
@edgarriba How about something like this? [Colab](https://colab.research.google.com/drive/1B09C0ZoOBdm4lB-6OwPfcztjIdlz5wd?usp=sharing)
